content/docs/pipelines/reference/component-spec.md

```diff
@@ -0,0 +1,135 @@
+    * `default`: Specifies the default value for an input. Only
+        valid for inputs.
+    * `type`: Specifies the type of input/output. The types are used
+        as hints for pipeline authors and can be used by the pipeline system/UI
+        to validate arguments and connections between components. Basic types
+        are **String**, **Integer**, **Number**, **Boolean**.
+
+### Implementation
+
+* `implementation`: Specifies how to execute the component instance.
+    There are two implementation types,  `container` and `graph`. (The latter is
+    not in scope for this document.) In future we may introduce more 
+    implementation types like `daemon` or `K8sResource`.
+
+    * `container`:
+        Describes the Docker container that implements the component. A portable 
+        subset of the Kubernetes
+        [Container v1 spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#container-v1-core).
+
+        * `image`: Name of the Docker image.
+        * `command`: Entrypoint array. The Docker image's
+            ENTRYPOINT is used if this is not provided. Each item is either a
+            string or a placeholder. The most common placeholders are
+            `{inputValue: Input name}` and `{outputPath: Output name}`.
+        * `args`: Arguments to the entrypoint. The Docker
+            image's CMD is used if this is not provided. Each item is either a
+            string or a placeholder. The most common placeholders are
+            `{inputValue: Input name}` and `{outputPath: Output name}`.
+        * `env`: Map of environment variables to set in the container.
+        * `fileOutputs`: Legacy property that is only needed in
+            cases where the container always stores the output data in some
+            hard-coded non-configurable local location. This property specifies
+            a map between some outputs and local file paths where the program
+            writes the output data files. Only needed for components that have
+            hard-coded output paths. Such containers need to be fixed by
+            modifying the program or adding a wrapper script that copies the
+            output to a configurable location. Otherwise the component may be
+            incompatible with future storage systems.
+
+You can set all other Kubernetes container properties when you
+use the component inside a pipeline.
+
+## Using placeholders for command-line arguments
+
+### Consuming input by value (parameter)
+
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
+
+### Outputs
+
+Output paths are filled in by the pipeline system. The Output construct is
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
+* In `component.yaml': `{outputPath: Trained model}`
+* Resulting Argo YAML: `"/outputs/trained_model/data"`
```

Comment:
is trained_model directory constructed from the "Trained model" name? In the docs - can we always keep the names matching? Or we need to reference a clear translation logic here.

Reply:
@Ark-kun

Reply:
>is trained_model directory constructed from the "Trained model" name?

Yes.

> can we always keep the names matching?

We'll fix this part to make the names match. This is only a path though and the user does not interact with it.

> Or we need to reference a clear translation logic here.

We can describe the translation logic. But half of the purpose of generating the paths (as opposed to hard-coding the paths) is that it's opaque and can be changed later. We could as well just generate UUID paths like `/outputs/5c9b2d651e6c01cb41d834faa77925157d5d1599/data`. The user should not really care about or depend on some concrete logic.

Reply:
Maybe we should not mention Argo here at all.

Reply:
I've added comments with replacement text. We should not mention Argo. (Remove this line)

Reply:
Done.